ldap/servers/slapd/back-ldbm/dblayer.c

```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
     return priv->dblayer_start_fn(li, dbmode);
 }
 
+/* Test or set instance import lock (shared among processes) */
+int
+dblayer_import_lock(backend *be, lock_mode_t mode)
+{
+    struct ldbminfo *li = (struct ldbminfo *)be->be_database->plg_private;
+    ldbm_instance *inst = be->be_instance_info;
+    char filename[MAXPATHLEN];
+    char *pt;
+    int len;
+    int fd;
+    int rc;
+
+    if (mode != LOCK_UNLOCK && inst->import_lock_fd >=0 ) {
+        /* Already locked in another thread */
+        return LDAP_BUSY;
+    }
+
+    pt = dblayer_get_full_inst_dir(li, inst, filename, MAXPATHLEN);
+        slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to get import lock filename\n");
+    if (pt != filename) {
+        slapi_ch_free_string(&pt);
+    }
+    if (!pt) {
+        slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to get import lock filename\n");
+        return SLAPI_FAIL_GENERAL;
+    }
+    len = strlen(filename);
+
+    if  (len + (sizeof IMPORT_LOCK_SUFFIX) >= MAXPATHLEN) {
+        slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to get import lock filename\n");
+        return SLAPI_FAIL_GENERAL;
+    }
+    memcpy(filename, IMPORT_LOCK_SUFFIX, (sizeof IMPORT_LOCK_SUFFIX));
+    switch (mode) {
+        default:
+            abort();
+
+        case LOCK_UNLOCK:
+            fd = inst->import_lock_fd;
+            if (fd>=0) {
+               /* Let's unlock other processes */
+               unlink(filename);
+               /* Let's unlock other threads */
+               inst->import_lock_fd = -1;
+               /* Lets release the resources */
+               flock(fd, LOCK_UN);
+               close(fd);
+            }
```

Comment:
indentation is off in the above code block.  :-)  There are more to come, sorry...